testnet/tools/nns-tools/add-release-to-changelog.sh

```diff
@@ -0,0 +1,149 @@
+    status.
+
+" >&2
+    exit 1
+}
+
+PROPOSAL_ID="${1}"
+
+cd $(repo_root)
+PWD=$(pwd)
+
+# Fetch the proposal.
+print_purple "Fetching proposal ${PROPOSAL_ID}..." >&2
+PROPOSAL_INFO=$(
+    __dfx --quiet \
+        canister call \
+        --ic \
+        --candid rs/nns/governance/canister/governance.did \
+        $(nns_canister_id governance) \
+        get_proposal_info "(${PROPOSAL_ID})" \
+        | idl2json
+)
+# Unwrap.
+LEN=$(echo "${PROPOSAL_INFO}" | jq '. | length')
+if [[ "${LEN}" -ne 1 ]]; then
+    print_red "Unexpected result from the get_proposal_info method:" >&2
+    print_red "Should have one element, but has ${LEN}" >&2
+    exit 1
+fi
+PROPOSAL_INFO=$(echo "${PROPOSAL_INFO}" | jq '.[0]')
+
+# Assert was executed.
+EXECUTED_TIMESTAMP_SECONDS=$(echo "${PROPOSAL_INFO}" | jq '.executed_timestamp_seconds | tonumber')
+if [[ "${EXECUTED_TIMESTAMP_SECONDS}" -eq 0 ]]; then
+    print_red "Proposal ${PROPOSAL_ID} exists, but was not successfully executed." >&2
+    exit 1
+fi
+SECONDS_AGO=$(($(date +%s) - "${EXECUTED_TIMESTAMP_SECONDS}"))
+EXECUTED_ON=$(
+    date --utc \
+        --date=@"${EXECUTED_TIMESTAMP_SECONDS}" \
+        --iso-8601
+)
+print_purple "Proposal ${PROPOSAL_ID} was executed ${SECONDS_AGO} seconds ago." >&2
+
+# Extract which canister was upgraded, and to what commit.
+TITLE=$(echo $PROPOSAL_INFO | jq -r '.proposal[0].summary' | head -n 1)
+CANISTER_NAME=$(
+    echo "${TITLE}" \
+        | sed 's/# Upgrade the //' | sed 's/ Canister to Commit .*//' \
+        | tr '[:upper:]' '[:lower:]'
+)
+DESTINATION_COMMIT_ID=$(echo "${TITLE}" | sed 's/# Upgrade the .* Canister to Commit //')
+
+# Fail if the proposal's commit is not checked out.
+if [[ $(git rev-parse HEAD) != $DESTINATION_COMMIT_ID* ]]; then
+    echo >&2
+    print_red "You currently have $(git rev-parse HEAD)" >&2
+    print_red "checked out, but this command only supports being run when" >&2
+    print_red "the proposal's commit (${DESTINATION_COMMIT_ID}) is checked out." >&2
+    exit 1
+fi
+
+# cd to the canister's primary code path.
+CANISTER_CODE_PATH=$(
+    get_nns_canister_code_location "${CANISTER_NAME}" \
+        | sed "s^${PWD}^.^g" \
+        | cut -d' ' -f1
+)
+cd "${CANISTER_CODE_PATH}"
+
+# Assert that there is a CHANGELOG.md file.
+if [[ ! -e CHANGELOG.md ]]; then
+    echo >&2
+    print_red "${CANISTER_NAME} has no CHANGELOG.md file." >&2
+    exit 1
+fi
+# TODO: Also verify that unreleased_changelog.md exists.
+# TODO: Verify that there are no uncommited changes in this dir, the canister's primary code path.
+
+# Construct new entry for CHANGELOG.md
+#
+# Python is used to filter out empty sections.
+# Python is used because I cannot figure out how to do that using mac sed.
+NEW_FEATURES_AND_FIXES=$(
+    sed '1,/^# Next Upgrade Proposal$/d' \
+        unreleased_changelog.md \
+        | python3 -c 'import sys, re; s = sys.stdin.read(); print(re.sub(r"## [\w ]+\n+(?=##|\Z)", "", s).strip())'
+)
```

Comment:
i would add some error handling for the python, otherwise I think it will fail silently if anything goes wrong, right?
```suggestion
PYTHON_SCRIPT='import sys, re; s = sys.stdin.read(); print(re.sub(r"## [\w ]+\n+(?=##|\Z)", "", s).strip())'
NEW_FEATURES_AND_FIXES=$(
    sed '1,/^# Next Upgrade Proposal$/d' \
        unreleased_changelog.md \
        | python3 -c "${PYTHON_SCRIPT}" 2>/tmp/py_error.$$
)
PYTHON_EXIT=$?

if [[ $PYTHON_EXIT -ne 0 ]]; then
    print_red "Python processing failed:" >&2
    cat /tmp/py_error.$$ >&2
    rm -f /tmp/py_error.$$
    exit 1
fi
rm -f /tmp/py_error.$$
```